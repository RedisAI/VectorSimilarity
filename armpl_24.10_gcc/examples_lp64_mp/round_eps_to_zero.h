#ifndef ROUND_EPS_TO_ZERO_H
#define ROUND_EPS_TO_ZERO_H

/*
 * Arm Performance Libraries version 24.10
 * SPDX-FileCopyrightText: Copyright 2015-2024 Arm Limited and/or its affiliates
 */

#include <float.h>
#include <math.h>

// On some architectures we occasionally see tiny differences between an expected
// answer of 0 and the calculated answer, so we choose not to display these
float round_eps_to_zero_f(float x) {
	return fabs(x) > FLT_EPSILON ? x : 0;
}
double round_eps_to_zero_d(double x) {
	return fabs(x) > DBL_EPSILON ? x : 0;
}

#endif
