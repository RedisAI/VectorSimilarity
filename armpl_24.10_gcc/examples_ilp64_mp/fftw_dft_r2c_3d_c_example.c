/*
 * fftw_dft_r2c_3d: FFT of a 3d real sequence
 *
 * Arm Performance Libraries version 24.10
 * SPDX-FileCopyrightText: Copyright 2015-2024 Arm Limited and/or its affiliates
 */

#include <complex.h>
#include <fftw3.h>
#include <math.h>
#include <stdio.h>

#include "round_eps_to_zero.h"

int main(void) {
#define MMAX 10
#define NMAX 10
#define LMAX 10
	double xx[MMAX * NMAX * LMAX];
	double x[MMAX * NMAX * LMAX];
	// The output vector is of size m * n * (l/2+1) as it is Hermitian
	fftw_complex y[MMAX * NMAX * (LMAX / 2 + 1)];

	printf("ARMPL example: FFT of a 3d real sequence using "
	       "fftw_plan_dft_r2c_3d\n");
	printf("-------------------------------------------------------------------"
	       "\n");
	printf("\n");

	/* The 3d sequence of double data */
	int m = 2;
	int n = 2;
	int l = 5;
	int ix = 0;
	int i, j, k;
	for (i = 1; i <= m; i++) {
		for (j = 1; j <= n; j++) {
			for (k = 1; k <= l; k++) {
				x[ix] = 1.0 + cos(ix + 1);
				xx[ix] = x[ix];
				ix++;
			}
		}
	}

	fftw_plan f = fftw_plan_dft_r2c_3d(m, n, l, x, y, FFTW_ESTIMATE);
	fftw_plan b = fftw_plan_dft_c2r_3d(m, n, l, y, x, FFTW_ESTIMATE);
	fftw_execute(f);
	fftw_destroy_plan(f);

	printf("Components of discrete Fourier transform:\n");
	printf("\n");
	ix = 0;
	for (i = 0; i < m; i++) {
		for (j = 0; j < n; j++) {
			for (k = 0; k <= l / 2; k++) {
				double scale = 1.0 / sqrt(m * n * l);
				double y_real = round_eps_to_zero_d(creal(y[ix]) * scale);
				double y_imag = round_eps_to_zero_d(cimag(y[ix]) * scale);
				printf("%4d,%d,%d   (%7.4f+%7.4fi)\n", i + 1, j + 1, k + 1, y_real, y_imag);
				ix++;
			}
		}
	}
	fftw_execute(b);
	fftw_destroy_plan(b);

	printf("\n");
	printf("Original sequence as restored by inverse transform:\n");
	printf("\n");
	printf("       Original  Restored\n");
	ix = 0;
	for (i = 0; i < m; i++) {
		for (j = 0; j < n; j++) {
			for (k = 0; k < l; k++) {
				double scale = 1.0 / (m * n * l);
				double xx_ix = round_eps_to_zero_d(xx[ix]);
				double x_ix = round_eps_to_zero_d(x[ix] * scale);
				printf("%4d,%d,%d   %7.4f   %7.4f\n", i + 1, j + 1, k + 1, xx_ix, x_ix);
				ix++;
			}
		}
	}

	return 0;
}
