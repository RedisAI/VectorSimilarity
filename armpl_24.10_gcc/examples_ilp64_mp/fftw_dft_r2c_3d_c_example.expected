ARMPL example: FFT of a 3d real sequence using fftw_plan_dft_r2c_3d
-------------------------------------------------------------------

Components of discrete Fourier transform:

   1,1,1   ( 4.5928+ 0.0000i)
   1,1,2   ( 0.1236+-0.5002i)
   1,1,3   ( 0.1220+-0.0530i)
   1,2,1   (-0.1667+ 0.0000i)
   1,2,2   ( 0.3302+ 0.0017i)
   1,2,3   ( 0.0563+ 0.0002i)
   2,1,1   (-0.7546+ 0.0000i)
   2,1,2   ( 1.4943+ 0.0075i)
   2,1,3   ( 0.2546+ 0.0008i)
   2,2,1   (-0.3047+ 0.0000i)
   2,2,2   (-0.3121+ 1.2632i)
   2,2,3   (-0.3080+ 0.1338i)

Original sequence as restored by inverse transform:

       Original  Restored
   1,1,1    1.5403    1.5403
   1,1,2    0.5839    0.5839
   1,1,3    0.0100    0.0100
   1,1,4    0.3464    0.3464
   1,1,5    1.2837    1.2837
   1,2,1    1.9602    1.9602
   1,2,2    1.7539    1.7539
   1,2,3    0.8545    0.8545
   1,2,4    0.0889    0.0889
   1,2,5    0.1609    0.1609
   2,1,1    1.0044    1.0044
   2,1,2    1.8439    1.8439
   2,1,3    1.9074    1.9074
   2,1,4    1.1367    1.1367
   2,1,5    0.2403    0.2403
   2,2,1    0.0423    0.0423
   2,2,2    0.7248    0.7248
   2,2,3    1.6603    1.6603
   2,2,4    1.9887    1.9887
   2,2,5    1.4081    1.4081
